include/envoy/stream_info/stream_info.h

```diff
@@ -453,6 +453,11 @@ class StreamInfo {
    */
   virtual Ssl::ConnectionInfoConstSharedPtr upstreamSslConnection() const PURE;
 
+  /**
+   * @param route_entry set for this request/connection.
+   */
+  virtual void setRouteEntry(const Router::RouteEntry* route_entry) PURE;
```

Comment:
I don't mind the idea of storing a route inside the stream info, but this will need to be a shared_ptr to a const object to avoid potential lifetime issues around RDS, etc.